cmd/registry/patch/artifact.go

```diff
@@ -60,69 +59,98 @@ func ExportArtifact(ctx context.Context, client *gapic.RegistryClient, message *
-	// https://stackoverflow.com/questions/66709979/dynamically-parse-yaml-field-to-one-of-a-finite-set-of-structs-in-go
-	type Alias Artifact
-	type Wrapper struct {
-		*Alias `yaml:",inline"`
-		Data   yaml.Node `yaml:"data"`
+// styleForYAML sets the style field on a tree of yaml.Nodes for JSON export.
+func styleForJSON(node *yaml.Node) {
+	switch node.Kind {
+	case yaml.DocumentNode, yaml.SequenceNode, yaml.MappingNode:
+		node.Style = yaml.FlowStyle
+	case yaml.ScalarNode:
+		switch node.Tag {
+		case "!!str":
+			node.Style = yaml.DoubleQuotedStyle
+		default:
+			node.Style = 0
+		}
+	case yaml.AliasNode:
+	default:
 	}
-	wrapper := &Wrapper{Alias: (*Alias)(a)}
-	if err := node.Decode(wrapper); err != nil {
-		return err
+	for _, n := range node.Content {
+		styleForJSON(n)
 	}
-	switch a.Kind {
-	case "DisplaySettings":
-		a.Data = new(DisplaySettingsData)
-	case "Lifecycle":
-		a.Data = new(LifecycleData)
-	case "Manifest":
-		a.Data = new(ManifestData)
-	case "ReferenceList":
-		a.Data = new(ReferenceListData)
-	case "TaxonomyList":
-		a.Data = new(TaxonomyListData)
-	default:
-		return fmt.Errorf("unable to unmarshal %s", a.Kind)
+}
+
+func removeIdAndKind(node *yaml.Node) *yaml.Node {
+	if node.Kind == yaml.MappingNode {
+		content := make([]*yaml.Node, 0)
+		for i := 0; i < len(node.Content); i += 2 {
+			k := node.Content[i]
+			if k.Value != "id" && k.Value != "kind" {
+				content = append(content, node.Content[i])
+				content = append(content, node.Content[i+1])
+			}
+		}
+		node.Content = content
 	}
-	return wrapper.Data.Decode(a.Data)
+	return node
 }
 
-func (a *Artifact) MarshalYAML() (interface{}, error) {
-	// This struct provides a temporary equivalent Artifact representation with
-	// a YAML struct tag that allows the Data field to be directly exported.
-	// The primary definition (above) has a dash ("-") for this field to defer
-	// unmarshalling until its type is known. But that causes YAML marshalling
-	// to skip the field, and rather than try to tweak the struct tags at
-	// runtime, we instead copy the artifact data into this equivalent
-	// structure that is tagged for direct export.
-	type exportable struct {
-		Header `yaml:",inline"`
-		Data   ArtifactData `yaml:"data"`
-	}
-	return &exportable{
-		Header: a.Header,
-		Data:   a.Data,
+func newArtifact(message *rpc.Artifact) (*Artifact, error) {
+	artifactName, err := names.ParseArtifact(message.Name)
+	if err != nil {
+		return nil, err
+	}
+	// Unmarshal the serialized protobuf containing the artifact content.
+	m, err := protoMessageForMimeType(message.MimeType)
+	if err != nil {
+		return nil, err
+	}
+	if err = proto.Unmarshal(message.Contents, m); err != nil {
+		return nil, err
+	}
+	// Marshal the artifact content as JSON using the protobuf marshaller.
+	var s []byte
+	s, err = protojson.MarshalOptions{
+		UseEnumNumbers:  false,
+		EmitUnpopulated: true,
+		Indent:          "  ",
+		UseProtoNames:   false,
```

Comment:
Are our protos still intended to be the source of truth for structure and documentation? If so, using the same field names seems like it would eliminate a lot of mental burden for users.

Reply:
Noted. I prefer this because I think there are consumption-side benefits to using protos (generated support code) and user-side benefits to following YAML conventions and that the camel-to-snake conversion isn't that hard. But I'm open to keeping an eye on this.